File: python/src/solana_keychain/privy/signer.py
"""Privy API signer integration."""

import base64
from dataclasses import dataclass, field
from typing import Any
from urllib.parse import quote

import httpx
from solders.pubkey import Pubkey
from solders.signature import Signature
from solders.transaction import Transaction

from solana_keychain.core.errors import SignerError, SignerErrorCode
from solana_keychain.core.http import assert_https_url, fetch_signer_json, normalize_base_url
from solana_keychain.core.signer import SignedTransaction, SolanaSigner
from solana_keychain.core.transaction_util import (
    add_signature_to_transaction,
    assert_unversioned_wire_transaction,
    classify_signed_transaction,
    get_signing_keypair_position,
    serialize_transaction,
)
from solana_keychain.privy.authorization import (
    DEFAULT_AUTHORIZATION_REQUEST_EXPIRY_MS,
    PrivyAuthorizationConfig,
    prepare_authorization_headers,
)

DEFAULT_API_BASE_URL = "https://api.privy.io/v1"


@dataclass
class PrivySignerConfig:
    """Configuration for a Privy signer.

    ``authorization_request_expiry_ms`` defaults to 15 minutes; set it to ``None``
    to omit ``privy-request-expiry`` from signed payloads and request headers.
    """

    app_id: str
    app_secret: str = field(repr=False)
    wallet_id: str
    api_base_url: str = DEFAULT_API_BASE_URL
    authorization_context: PrivyAuthorizationConfig | None = field(default=None, repr=False)
    authorization_request_expiry_ms: int | None = DEFAULT_AUTHORIZATION_REQUEST_EXPIRY_MS
    http_client: httpx.AsyncClient | None = field(default=None, repr=False)


class PrivySigner(SolanaSigner):
    """Signer backed by a Privy-held wallet.

    ``init()`` must be awaited before use — it resolves the wallet's public key.
    ``create_privy_signer()`` does this for you.
    """

    def __init__(self, config: PrivySignerConfig) -> None:
        api_base_url = normalize_base_url(config.api_base_url)
        assert_https_url(api_base_url, "api_base_url")
        self._api_base_url = api_base_url
        self._app_id = config.app_id
        self._app_secret = config.app_secret
        self._wallet_id = config.wallet_id
        self._authorization_context = config.authorization_context
        self._authorization_request_expiry_ms = config.authorization_request_expiry_ms
        self._http_client = config.http_client
        self._public_key: Pubkey | None = None

    def __repr__(self) -> str:
        return f"PrivySigner(pubkey={self._public_key})"

    async def init(self) -> None:
        """Resolve the wallet's public key. Must be awaited before signing."""
        self._public_key = await self._fetch_public_key()

    def _initialized_pubkey(self) -> Pubkey:
        if self._public_key is None:
            raise SignerError(
                SignerErrorCode.NOT_INITIALIZED,
                "PrivySigner is not initialized; call init() before signing",
            )
        return self._public_key

    @property
    def pubkey(self) -> Pubkey:
        return self._initialized_pubkey()

    def _base_headers(self) -> dict[str, str]:
        credentials = base64.b64encode(f"{self._app_id}:{self._app_secret}".encode()).decode(
            "ascii"
        )
        return {"Authorization": f"Basic {credentials}", "privy-app-id": self._app_id}

    async def _fetch_public_key(self) -> Pubkey:
        url = f"{self._api_base_url}/wallets/{quote(self._wallet_id, safe='')}"
        wallet = await fetch_signer_json(
            url=url,
            provider_name="Privy",
            headers=self._base_headers(),
            client=self._http_client,
        )
        if not isinstance(wallet, dict):
            raise SignerError(SignerErrorCode.REMOTE_API_ERROR, "Invalid Privy wallet response")
        chain_type = wallet.get("chain_type")
        if chain_type != "solana":
            raise SignerError(
                SignerErrorCode.REMOTE_API_ERROR,
                f"Expected Solana wallet, got chain_type={chain_type}",
            )
        try:
            return Pubkey.from_string(wallet.get("address", ""))
        except Exception:
            raise SignerError(
                SignerErrorCode.INVALID_PUBLIC_KEY, "Invalid public key from Privy API"
            ) from None

    async def _post_rpc(self, request: dict[str, Any]) -> Any:
        url = f"{self._api_base_url}/wallets/{quote(self._wallet_id, safe='')}/rpc"
        authorization_signature, request_expiry = prepare_authorization_headers(
            app_id=self._app_id,
            authorization_config=self._authorization_context,
            method="POST",
            url=url,
            body=request,
            request_expiry_ms=self._authorization_request_expiry_ms,
        )
        headers = self._base_headers()
        headers["Content-Type"] = "application/json"
        if authorization_signature is not None:
            headers["privy-authorization-signature"] = authorization_signature
        if request_expiry is not None:
            headers["privy-request-expiry"] = request_expiry

        return await fetch_signer_json(
            url=url,
            provider_name="Privy",
            method="POST",
            headers=headers,
            json_body=request,
            client=self._http_client,
        )

    async def _sign_bytes(self, message: bytes) -> Signature:
        public_key = self._initialized_pubkey()
        request = {
            "method": "signMessage",
            "chain_type": "solana",
            "params": {"message": base64.b64encode(message).decode("ascii"), "encoding": "base64"},
        }
        response = await self._post_rpc(request)

        data = response.get("data") if isinstance(response, dict) else None
        signature_b64 = data.get("signature") if isinstance(data, dict) else None
        if not isinstance(signature_b64, str):
            raise SignerError(SignerErrorCode.REMOTE_API_ERROR, "No signature in Privy response")
        try:
            signature_bytes = base64.b64decode(signature_b64, validate=True)
        except Exception:
            raise SignerError(
                SignerErrorCode.SERIALIZATION_ERROR,
                "Failed to decode base64 signature from Privy",
            ) from None
        try:
            signature = Signature.from_bytes(signature_bytes)
        except Exception:
            raise SignerError(SignerErrorCode.SIGNING_FAILED, "Failed to parse signature") from None
        if not signature.verify(public_key, message):
            raise SignerError(
                SignerErrorCode.SIGNING_FAILED,
                "Signature verification failed — the returned signature does not match "
                "the public key",
            )
        return signature

    async def sign_transaction(self, transaction: Transaction) -> SignedTransaction:
        """Sign via Privy's ``signTransaction`` RPC, submitting the full wire
        transaction so wallet policies with transaction conditions apply.
        Policies must allow the ``signTransaction`` method."""
        public_key = self._initialized_pubkey()
        request = {
            "method": "signTransaction",
            "chain_type": "solana",
            "params": {
                "transaction": base64.b64encode(bytes(transaction)).decode("ascii"),
                "encoding": "base64",
            },
        }
        response = await self._post_rpc(request)

        data = response.get("data") if isinstance(response, dict) else None
        signed_b64 = data.get("signed_transaction") if isinstance(data, dict) else None
        if not isinstance(signed_b64, str):
            raise SignerError(
                SignerErrorCode.REMOTE_API_ERROR, "No signed_transaction in Privy response"
            )
        try:
            signed_wire = base64.b64decode(signed_b64, validate=True)
        except Exception:
            raise SignerError(
                SignerErrorCode.SERIALIZATION_ERROR,
                "Failed to decode signed transaction returned by Privy",
            ) from None
        assert_unversioned_wire_transaction("Privy", signed_wire)
        try:
            signed = Transaction.from_bytes(signed_wire)
        except Exception:
            raise SignerError(
                SignerErrorCode.SERIALIZATION_ERROR,
                "Failed to decode signed transaction returned by Privy",
            ) from None

        position = get_signing_keypair_position(signed, public_key)
        signatures = signed.signatures
        if position >= len(signatures):
            raise SignerError(
                SignerErrorCode.SIGNING_FAILED,
                "Privy signature slot missing from returned transaction",
            )
        signature = signatures[position]
        if not signature.verify(public_key, transaction.message_data()):
            raise SignerError(
                SignerErrorCode.SIGNING_FAILED,
                "Signature verification failed — the returned signature does not match "
                "the public key",
            )
        add_signature_to_transaction(transaction, public_key, signature)
        return classify_signed_transaction(
            transaction, serialize_transaction(transaction), signature
        )

    async def sign_message(self, message: bytes) -> Signature:
        return await self._sign_bytes(message)

    async def is_available(self) -> bool:
        if self._public_key is None:
            return False
        try:
            return await self._fetch_public_key() == self._public_key
        except SignerError:
            return False


async def create_privy_signer(config: PrivySignerConfig) -> PrivySigner:
    """Create a ready-to-use Privy signer (awaits ``init()``)."""
    signer = PrivySigner(config)
    await signer.init()
    return signer

File: go/signers/privy/signer.go
package privy

import (
	"bytes"
	"context"
	"encoding/base64"
	"encoding/json"
	"net/http"
	"strconv"

	"github.com/gagliardetto/solana-go"

	"github.com/solana-foundation/solana-keychain/go/core"
)

// Signer signs with a Privy wallet via Privy's REST API. All fields are
// immutable after New, so a Signer is safe for concurrent use.
type Signer struct {
	appID                        string
	appSecret                    string
	walletID                     string
	apiBaseURL                   string
	client                       *http.Client
	pubkey                       solana.PublicKey
	authorizationContext         *AuthorizationContext
	authorizationRequestExpiryMs *uint64
}

// Ensure Signer satisfies the core contract at compile time.
var _ core.Signer = (*Signer)(nil)

// New builds a Privy signer and fetches the wallet's public key from the Privy
// API, so the returned signer is ready to use.
func New(ctx context.Context, cfg Config) (*Signer, error) {
	if cfg.AppID == "" || cfg.AppSecret == "" || cfg.WalletID == "" {
		return nil, core.NewSignerError(core.CodeConfigError,
			"missing required configuration fields (AppID, AppSecret, or WalletID)")
	}
	client := cfg.HTTPClient
	if client == nil {
		client = core.NewHTTPClient(cfg.HTTPClientConfig)
	}
	baseURL, err := core.NormalizeHTTPSBaseURL(cfg.APIBaseURL, DefaultAPIBaseURL, "api_base_url")
	if err != nil {
		return nil, err
	}
	s := &Signer{
		appID:                        cfg.AppID,
		appSecret:                    cfg.AppSecret,
		walletID:                     cfg.WalletID,
		apiBaseURL:                   baseURL,
		client:                       client,
		authorizationContext:         cfg.AuthorizationContext,
		authorizationRequestExpiryMs: resolveAuthorizationRequestExpiryMs(cfg),
	}
	pubkey, err := s.fetchPublicKey(ctx)
	if err != nil {
		return nil, err
	}
	s.pubkey = pubkey
	return s, nil
}

// Pubkey returns the wallet's public key fetched during New.
func (s *Signer) Pubkey() solana.PublicKey { return s.pubkey }

// String renders the signer without any secret material.
func (s Signer) String() string {
	return "privy.Signer{pubkey: " + s.pubkey.String() + ", apiBaseURL: " + s.apiBaseURL + "}"
}

// GoString mirrors String so %#v cannot leak secrets either.
func (s Signer) GoString() string { return s.String() }

// SignMessage signs arbitrary bytes with the Privy wallet and verifies the
// returned signature against the wallet's public key before returning it.
func (s *Signer) SignMessage(ctx context.Context, message []byte) (solana.Signature, error) {
	return s.signBytes(ctx, message)
}

// SignTransaction signs tx via Privy's signTransaction RPC, submitting the
// full wire transaction so wallet policies with transaction conditions apply.
// Policies must allow the signTransaction method.
func (s *Signer) SignTransaction(ctx context.Context, tx *solana.Transaction) (core.SignedTransaction, error) {
	msg, err := tx.Message.MarshalBinary()
	if err != nil {
		return core.SignedTransaction{}, core.WrapSignerError(core.CodeSerializationError, "failed to serialize transaction message", err)
	}
	unsignedWire, err := tx.MarshalBinary()
	if err != nil {
		return core.SignedTransaction{}, core.WrapSignerError(core.CodeSerializationError, "failed to serialize transaction", err)
	}

	request := signTransactionRequest{
		Method:    "signTransaction",
		ChainType: "solana",
		Params: signTransactionParams{
			Transaction: base64.StdEncoding.EncodeToString(unsignedWire),
			Encoding:    "base64",
		},
	}
	body, err := s.postRPC(ctx, request)
	if err != nil {
		return core.SignedTransaction{}, err
	}

	var signResp signTransactionResponse
	if err := json.Unmarshal(body, &signResp); err != nil {
		return core.SignedTransaction{}, core.WrapSignerError(core.CodeSerializationError, "failed to parse privy signing response", err)
	}
	signedWire, err := base64.StdEncoding.DecodeString(signResp.Data.SignedTransaction)
	if err != nil {
		return core.SignedTransaction{}, core.NewSignerError(core.CodeSerializationError,
			"failed to decode signed transaction returned by privy")
	}
	returned, err := solana.TransactionFromBytes(signedWire)
	if err != nil {
		return core.SignedTransaction{}, core.WrapSignerError(core.CodeSerializationError,
			"failed to deserialize signed transaction returned by privy", err)
	}

	position, err := core.SigningPosition(returned, s.pubkey)
	if err != nil {
		return core.SignedTransaction{}, err
	}
	if position >= len(returned.Signatures) {
		return core.SignedTransaction{}, core.NewSignerError(core.CodeSigningFailed,
			"privy signature slot missing from returned transaction")
	}
	sig := returned.Signatures[position]
	if err := core.VerifySignature(s.pubkey, msg, sig); err != nil {
		return core.SignedTransaction{}, err
	}

	return core.AttachSignature(tx, s.pubkey, sig)
}

// postRPC sends a wallet RPC request with Privy auth and
// authorization-signature headers and returns the response body on 2xx.
func (s *Signer) postRPC(ctx context.Context, request any) ([]byte, error) {
	url := s.apiBaseURL + "/wallets/" + s.walletID + "/rpc"
	reqBody, err := json.Marshal(request)
	if err != nil {
		return nil, core.WrapSignerError(core.CodeSerializationError, "failed to serialize privy signing request", err)
	}
	authHeaders, err := s.prepareAuthorizationHeaders(http.MethodPost, url, request)
	if err != nil {
		return nil, err
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader(reqBody))
	if err != nil {
		return nil, core.WrapSignerError(core.CodeHTTPError, "failed to build privy signing request", err)
	}
	req.Header.Set("Authorization", s.authHeader())
	req.Header.Set("privy-app-id", s.appID)
	req.Header.Set("Content-Type", "application/json")
	if authHeaders.signature != "" {
		req.Header.Set("privy-authorization-signature", authHeaders.signature)
	}
	if authHeaders.requestExpiry != "" {
		req.Header.Set("privy-request-expiry", authHeaders.requestExpiry)
	}
	return s.do(req, "privy signing")
}

// IsAvailable re-fetches the wallet from the Privy API and reports whether it
// still resolves to this signer's public key. All errors are swallowed and
// reported as false.
func (s *Signer) IsAvailable(ctx context.Context) bool {
	ctx, cancel := context.WithTimeout(ctx, core.AvailabilityTimeout)
	defer cancel()
	pubkey, err := s.fetchPublicKey(ctx)
	return err == nil && pubkey == s.pubkey
}

// authHeader returns the HTTP Basic auth header value for the app credentials.
func (s *Signer) authHeader() string {
	return "Basic " + base64.StdEncoding.EncodeToString([]byte(s.appID+":"+s.appSecret))
}

// fetchPublicKey resolves the wallet's Solana address via GET /wallets/{id}.
func (s *Signer) fetchPublicKey(ctx context.Context) (solana.PublicKey, error) {
	url := s.apiBaseURL + "/wallets/" + s.walletID
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return solana.PublicKey{}, core.WrapSignerError(core.CodeHTTPError, "failed to build privy wallet request", err)
	}
	req.Header.Set("Authorization", s.authHeader())
	req.Header.Set("privy-app-id", s.appID)

	body, err := s.do(req, "privy wallet fetch")
	if err != nil {
		return solana.PublicKey{}, err
	}

	var wallet walletResponse
	if err := json.Unmarshal(body, &wallet); err != nil {
		return solana.PublicKey{}, core.WrapSignerError(core.CodeSerializationError, "failed to parse privy wallet response", err)
	}

	if wallet.Address == "" {
		return solana.PublicKey{}, core.NewSignerError(core.CodeRemoteAPIError, "missing address in privy wallet response")
	}
	if wallet.ChainType != "solana" {
		return solana.PublicKey{}, core.NewSignerError(core.CodeRemoteAPIError, "expected Solana wallet, got chain_type="+wallet.ChainType)
	}

	// For Solana wallets, the address is the public key.
	pubkey, err := solana.PublicKeyFromBase58(wallet.Address)
	if err != nil {
		return solana.PublicKey{}, core.NewSignerError(core.CodeInvalidPublicKey, "invalid public key from privy API")
	}
	return pubkey, nil
}

// signBytes signs message via POST /wallets/{id}/rpc with method "signMessage",
// sending the bytes base64-encoded and decoding the base64 signature from the
// response, then verifies it locally.
func (s *Signer) signBytes(ctx context.Context, message []byte) (solana.Signature, error) {
	request := signMessageRequest{
		Method:    "signMessage",
		ChainType: "solana",
		Params: signMessageParams{
			Message:  base64.StdEncoding.EncodeToString(message),
			Encoding: "base64",
		},
	}
	body, err := s.postRPC(ctx, request)
	if err != nil {
		return solana.Signature{}, err
	}

	var signResp signMessageResponse
	if err := json.Unmarshal(body, &signResp); err != nil {
		return solana.Signature{}, core.WrapSignerError(core.CodeSerializationError, "failed to parse privy signing response", err)
	}

	sig, err := core.DecodeSignatureBase64(signResp.Data.Signature, "privy")
	if err != nil {
		return solana.Signature{}, err
	}

	if err := core.VerifySignature(s.pubkey, message, sig); err != nil {
		return solana.Signature{}, err
	}
	return sig, nil
}

// do executes the request and returns the response body, mapping transport
// failures to CodeHTTPError and non-2xx statuses to CodeRemoteAPIError whose
// detail carries only the status code ("API error {status}").
func (s *Signer) do(req *http.Request, what string) ([]byte, error) {
	status, body, err := core.SendRequest(s.client, req, "privy")
	if err != nil {
		return nil, err
	}
	if !core.IsSuccess(status) {
		// Only the status code goes into the detail; the response body is
		// deliberately discarded since it may echo request material.
		return nil, core.NewSignerError(core.CodeRemoteAPIError,
			what+" failed: API error "+strconv.Itoa(status))
	}
	return body, nil
}
